pkg/webhooks/cluster_webhook.go

```diff
@@ -0,0 +1,204 @@
+	var allErrs field.ErrorList
+	if in.Spec.InfraType == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "infraType"), "must be set"))
+	}
+
+	if in.Spec.Region == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "region"), "must be set"))
+	}
+
+	if in.Spec.Credential == nil || in.Spec.Credential.SecretRef == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "credential", "secretRef"), "must be set"))
+	}
+
+	return allErrs
+}
+
+func validateVersion(in *v1.Cluster) field.ErrorList {
+	var allErrs field.ErrorList
+	if in.Spec.Version == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "version"), "must be set"))
+	} else {
+		if !version.KubeSemver.MatchString(in.Spec.Version) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "version"), in.Spec.Version, "must be a valid Kubernetes version"))
+		}
+	}
+
+	return allErrs
+}
+
+func validateNetwork(network v1.NetworkConfig, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateCIDR(network.VPC.CIDRBlock, fldPath.Child("vpc", "cidrBlock"))...)
+	allErrs = append(allErrs, validateCIDRBlocks(network.PodCIDRs, fldPath.Child("podCIDRs"))...)
+	allErrs = append(allErrs, validateCIDRBlocks(network.ServiceCIDRs, fldPath.Child("serviceCIDRs"))...)
+
+	return allErrs
+}
+
+func validateCIDR(cidr string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if _, _, err := net.ParseCIDR(cidr); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, cidr, err.Error()))
+	}
+	return allErrs
+}
+
+func validateCIDRBlocks(cidrBlocks []string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	for i, cidr := range cidrBlocks {
+		if _, _, err := net.ParseCIDR(cidr); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i), cidr, err.Error()))
+		}
+	}
+
+	return allErrs
+}
+
+func validateAdditionalResource(in []v1.ResourceRef, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(in) == 0 {
+		return allErrs
+	}
+
+	keys := sets.NewString()
+	for i, resource := range in {
+		if resource.Name == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Index(i).Child("name"), "must be set"))
+		}
+		if resource.Kind != "Secret" && resource.Kind != "ConfigMap" {
+			allErrs = append(allErrs, field.Required(fldPath.Index(i).Child("type"), "must be either Secret or ConfigMap"))
+		}
+		key := fmt.Sprintf("%s/%s", resource.Kind, resource.Name)
+		if keys.Has(key) {
+			allErrs = append(allErrs, field.Duplicate(fldPath.Index(i), key))
+		} else {
+			keys.Insert(key)
+		}
+	}
+
+	return allErrs
+}
+
+func (wh *ClusterWebhook) validate(in *v1.Cluster) error {
+	var allErrs field.ErrorList
+
+	// The Cluste name is used as a label value, so it must be a valid label value.
+	if errs := validation.IsValidLabelValue(in.Name); len(errs) > 0 {
```

Comment:
I think this is not user friendly

Reply:
Cluster API require this, we can do some conversion?